### ISSUE TITLE
segfault in atomic data sub when trying to interpolate collision strengths

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#938]
I'm getting a segfault in the first ionization cycle when running with the default behaviour on python 86, commit 5966ba8b8bec95277916891ff0361468cf8086e2. The parameter file is attached. I've tried running with and without ```-include_partial_cells```. With the flag, the code runs fine, without I get a segmentation fault: 

```
!!python: Total photon luminosity before transphot 3.028504726291e+46
xxx Mon Jan 31 15:24:22 2022    135.6 NOK                  Photon transport started

 Ion. Cycle 1/20 of test : Photon          0 of      17857 or    0.0 per cent 
[cpu-p-110:233630] *** Process received signal ***
[cpu-p-110:233630] Signal: Segmentation fault (11)
[cpu-p-110:233630] Signal code: Address not mapped (1)
[cpu-p-110:233630] Failing at address: 0x4a31ab905940
[cpu-p-110:233630] [ 0] /lib64/libpthread.so.0(+0xf630)[0x2b6723e6f630]
[cpu-p-110:233630] [ 1] /home/jhm70/python/bin/py[0x41cafe]
[cpu-p-110:233630] [ 2] /home/jhm70/python/bin/py[0x41bd75]
[cpu-p-110:233630] [ 3] /home/jhm70/python/bin/py[0x40d559]
[cpu-p-110:233630] [ 4] /home/jhm70/python/bin/py[0x40d9a8]
[cpu-p-110:233630] [ 5] /home/jhm70/python/bin/py[0x419809]
[cpu-p-110:233630] [ 6] /home/jhm70/python/bin/py[0x41a10e]
[cpu-p-110:233630] [ 7] /home/jhm70/python/bin/py[0x46c43e]
[cpu-p-110:233630] [ 8] /home/jhm70/python/bin/py[0x4050b9]
[cpu-p-110:233630] [ 9] /lib64/libc.so.6(__libc_start_main+0xf5)[0x2b672409e555]
[cpu-p-110:233630] [10] /home/jhm70/python/bin/py[0x405507]
[cpu-p-110:233630] *** End of error message ***
```
I've tried with a few different versions of MPI/gcc, this particular failure was with 

```
$ mpirun --version
mpirun (Open MPI) 4.0.4

$ mpicc --version
gcc (Spack GCC) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by jhmatthews** [comment]
Here's the pf file, sorry 
[qso_segfault_macro.pf.txt](https://github.com/agnwinds/python/files/7971986/qso_segfault_macro.pf.txt)

**Comment by jhmatthews** [comment]
This segfault is still happening in certain situations. Seems a bit compiler dependent.

**Comment by jhmatthews** [comment]
Possible this is related to issue #928 which also gets a segfault.

**Comment by jhmatthews** [comment]
Ok, so the original problem may have been related to the current problem, but seemed to go away when we recompiled with consistent compilers for libraries. Now we are getting a segfault in runs l

The statement of the current problem is that I get a segfault when trying to populate the macro-atom rate matrices - the traceback looks like this: 
```
==== backtrace (tid: 109277) ====
 0 0x000000000040ff0b linterp()  ???:0
 1 0x000000000040fd4f q21()  ???:0
 2 0x00000000004a6897 calc_all_matom_matrices()  ???:0
 3 0x000000000049aa79 calculate_ionization()  ???:0
 4 0x000000000040567f main()  ???:0
 5 0x0000000000022555 __libc_start_main()  ???:0
 6 0x0000000000404b69 _start()  ???:0
```

with some additional diagnostic work the call to q21 is happening in the simple atom section, around line 533 in macro_accelerate.c - the segfault actually happens at the call to fraction(), which is called by linterp() which is called by upsilon(), I believe. This might suggest an issue with the collision strength data, or it could be this is a red herring.

Pf file that creates the problem (note only segfaults on the cluster for me, so hard to diagnose): [run14_thmin45_rv1e19_fill0p01_massrate5p0.pf.txt](https://github.com/agnwinds/python/files/8161108/run14_thmin45_rv1e19_fill0p01_massrate5p0.pf.txt)

**Comment by jhmatthews** [comment]
I think I may have solved this - in atomic.h, we set the n_points member of the coll_stren structure to be a double:

```
typedef struct coll_stren
{
  int n;                        //Internal index
  int lower;                    //The lower energy level - this is in Chianti notation and is currently unused
  int upper;                    //The upper energy level - this is in Chianti notation and is currently unused
  double energy;                //The energy of the transition
  double gf;                    //The effective oscillator strength - NSH thinks this is oscillator strtength x lower level multiplicity
  double hi_t_lim;              //The high temerature limit
  double n_points;              //The number of points in the splie fit
  int type;                     //The type of fit, this defines how one computes the scaled temperature and scaled coll strength
  float scaling_param;          //The scaling parameter C used in the Burgess and Tully calculations
  double sct[N_COLL_STREN_PTS]; //The scaled temperature points in the fit
  double scups[N_COLL_STREN_PTS];       //The sclaed coll sttengths in ythe fit.
} Coll_stren, *Coll_strenptr;

extern Coll_stren coll_stren[NLINES];  
```
if I change the n_points variable to an int, the problem seems to go away. The n_points variable is passed to linterp which expects an int (I don't know why the compiler doesn't complain here). I'm assuming that something about the icc compiler on the cluster does something different to gcc or clang when this happens, and results in strange behaviour when trying to index an array. I don't know whether the gcc/clang behaviour will produce correct results or not, but I'll check this.

**Comment by jhmatthews** [comment]
I've checked that on my mac, with clang, changing the n_points between double and int has no effect on the calculated value of q21. Suggests to me that probably most compilers do this conversion OK and the icc ones are just different, so I guess we can close this issue once the variable type has been changed in dev.

@Higginbottom I'm assuming this is just a typo and there's not a reason for having double n_points?
